### DISCUSSION TITLE
Some improvements for records with ToString() and formatting

### DISCUSSION THREAD

**Discussion by Sergio0694** [dotnet/roslyn#52975] [Ideas]
> Opening this as a follow up from a conversation with @333fred on the C# Discord server.

This discussion is about a couple ideas for improvements to records in C# 10, both for performance and for better interoperability with the new [`InterpolatedStringBuilder` type](https://github.com/dotnet/runtime/issues/50601) in .NET 6. Specifically, it pertains to some possible optimizations that could be done to the generated `ToString()` method for records, as well as more efficient support for formatting.

### Faster `ToString()` method

If we use as an example a record such as `record Person(string FirstName, string LastName, int Age)`, we get this ([sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+tJQCYAIAKMUAzhAHYAUAAgIwAMuAYgJYkAuAcgIYC2MANLloMAMl2KdeA3MzJtcAQQDmMAJQBuIA==)):

```csharp
public override string ToString()
{
    StringBuilder stringBuilder = new StringBuilder();
    stringBuilder.Append("Person");
    stringBuilder.Append(" { ");
    if (PrintMembers(stringBuilder))
    {
        stringBuilder.Append(" ");
    }
    stringBuilder.Append("}");
    return stringBuilder.ToString();
}

protected virtual bool PrintMembers(StringBuilder builder)
{
    builder.Append("FirstName");
    builder.Append(" = ");
    builder.Append((object)FirstName);
    builder.Append(", ");
    builder.Append("LastName");
    builder.Append(" = ");
    builder.Append((object)LastName);
    builder.Append(", ");
    builder.Append("Age");
    builder.Append(" = ");
    builder.Append(Age.ToString());
    return true;
}
```

This is not efficient, both because `StringBuilder` doesn't pool the underlying buffer, and for all the various other disadvantages also listed [here](https://github.com/dotnet/csharplang/issues/2302). It would be much more efficient if the generated code for `ToString()` on C# 10 leveraged `InterpolatedStringBuilder` instead, to get all the benefits such as reduced allocations, no boxing, pooled underlying buffers, etc. I understand that this would also mean that `PrintMembers` could no longer be used (as it just takes a `StringBuilder`), so there might be a small code size increase in `ToString()` that might or might not be a concern, but worth pointing out.

### Better support for formatting

In the context of the improved string interpolation support in C# 10, I think it'd make sense if records in C# 10 also got an autogenerated `ISpanFormattable` implementation, or some equivalent mechanism to allow them to be used into interpolated string holes with much better efficiency. Having them implement `ISpanFormattable` seems like the most straightforward option, as `InterpolatedStringBuilder` and other types already have explicit support for this interface anyway.

Quoting Fred here:

> "Just make records better when used in formatting"

Cheers! 😄

**Comment by jcouv** [comment]
This was discussed in LDM 5/3/2021. See upcoming notes.

In short, the `ToString` implementation on records is meant to be minimal (no localization) and not performance-oriented. We recommend users implement `ISpanFormattable` and other `ToString` overload themselves if they want. Most likely the default format would not be good for such scenarios anyways...

**Comment by alrz** [comment]
I wish we could remove that `if` by moving the space to PrintMembers (if appropriate). Not exactly sure if that could work but the current codegen is kinda clunky.

Also no need to load builder in each line. just use Append return value?

**Reply by jcouv** [reply]
In order to avoid printing double spaces or trailing comma, some boolean needs to be passed around, but as far as I know, it cannot be removed.

As for chaining the `Append` calls, that sounds like a good idea.

**Reply by alrz** [reply]
What I'm suggesting is to replace `return true` with `Append(" ")`. Need to think more why that might not work. But that way we could chain Append calls all the way.
